Samples/Use multiple voucher codes on one order/Resources/Litium.Accelerator/Services/CheckoutServiceImpl.cs

```diff
@@ -0,0 +1,824 @@
+            if (_requestModelAccessor.RequestModel.Cart.OrderCarrier == null || !IsOrderPlaced((OrderState)_requestModelAccessor.RequestModel.Cart.OrderCarrier.OrderStatus))
+            {
+                throw new Exception(CheckoutConstants.StringPaymentUnsuccessful.AsWebSiteString());
+            }
+
+            return GetOrderConfirmationPageUrl(order);
+        }
+
+        private bool IsOrderPlaced(OrderState orderState)
+        {
+            return orderState == OrderState.WaitingConfirmation
+                   || orderState == OrderState.Confirmed
+                   || orderState == OrderState.Processing
+                   || orderState == OrderState.Completed;
+        }
+
+        public override void ChangePaymentMethod(string paymentMethodId)
+        {
+            var currentLanguage = _languageService.Get(_requestModelAccessor.RequestModel.ChannelModel.Channel.WebsiteLanguageSystemId.Value);
+            var checkoutFlowInfo = _requestModelAccessor.RequestModel.Cart.CheckoutFlowInfo;
+            checkoutFlowInfo.SetValue(CheckoutConstants.ClientLanguage, currentLanguage.CultureInfo.Name);
+            checkoutFlowInfo.SetValue(CheckoutConstants.ClientTwoLetterISOLanguageName, currentLanguage.CultureInfo.TwoLetterISOLanguageName);
+
+            var currentOrderCarrier = _requestModelAccessor.RequestModel.Cart.OrderCarrier;
+            var paymentMethodParts = paymentMethodId.Split(':');
+            var paymentMethod = _moduleECommerce.PaymentMethods.Get(paymentMethodParts[1], paymentMethodParts[0], SecurityToken.CurrentSecurityToken);
+
+            var existingPaymentInfo = currentOrderCarrier.PaymentInfo.Any(x => x.PaymentProvider == paymentMethod.PaymentProviderName && x.PaymentMethod == paymentMethod.Name);
+            _moduleECommerce.CheckoutFlow.AddPaymentInfo(currentOrderCarrier, paymentMethod.PaymentProviderName, paymentMethod.Name, new AddressCarrier(), SecurityToken.CurrentSecurityToken);
+
+            if (!existingPaymentInfo)
+            {
+                var currentPaymentInfo = currentOrderCarrier.PaymentInfo.First(x => x.PaymentProvider == paymentMethod.PaymentProviderName && x.PaymentMethod == paymentMethod.Name);
+                currentPaymentInfo.TransactionNumber = null;
+                currentPaymentInfo.TransactionReference = null;
+            }
+
+            _moduleECommerce.Orders.CalculateOrderTotals(currentOrderCarrier, SecurityToken.CurrentSecurityToken);
+        }
+
+        public override void ChangeDeliveryMethod(Guid deliveryMethodId)
+        {
+            var currentOrderCarrier = _requestModelAccessor.RequestModel.Cart.OrderCarrier;
+            _moduleECommerce.CheckoutFlow.AddDelivery(currentOrderCarrier, deliveryMethodId, new AddressCarrier(), SecurityToken.CurrentSecurityToken);
+            _moduleECommerce.Orders.CalculateOrderTotals(currentOrderCarrier, SecurityToken.CurrentSecurityToken);
+        }
+
+        public override bool SetCampaignCode(string campaignCode)
```

Comment:
It's hard to follow exactly what you have added/changed in this file but I think it is in this method. Can we rewrite all this file into a decorator of the service instead and only override this method?